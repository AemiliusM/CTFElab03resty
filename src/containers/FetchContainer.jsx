import React, { Component } from "react";
import Controls from "../components/fetches/Controls-Methods";
import HistoryList from "../components/fetches/HistoryList";
import FetchRes from "../components/fetches/Response";
import { fetchMethods } from "../services/fetchmethods";

export default class FetchContainer extends Component {
    state = {
        loading: false,
        method: '',
        url: '',
        jsonValue: '',
        history: [],
        response: [],
    };

    handleInputChange = (event) => {
        const { name, value } = event.target;
        
        this.setState({
            [name]: value
        });
    }

    handleSubmit = async (event) => {
        event.preventDefault();

        this.setState({ loading: true})

        const { method, url, jsonValue, history } = this.state;
        let response;

        response = await fetchMethods(method, url, jsonValue);

        this.setState({
            loading: false, response,
            history: history.includes(url)
            ? history : [...history, { method: method, url: url }],
        });

    }

    render() {
        const {loading, url, method, jsonValue, history, response} = this.state;
        console.log('HISTORY', history)
        return (
            <>
            {loading ? (
                <h1>Loading Results</h1>
            ) : (
                <>
                <Controls
                url={url}
                method={method}
                jsonValue={jsonValue}
                onSubmit={this.handleSubmit}
                onInputChange={this.handleInputChange}
                />
                <HistoryList history={history} />
                <FetchRes response={response} />
                </>
            )}
            </>
        )
    }

}