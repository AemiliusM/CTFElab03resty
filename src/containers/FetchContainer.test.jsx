import { render, screen, waitFor } from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import React from "react";
import FetchContainer from "./FetchContainer";

describe('shows fetch tool and results', () => {
    it('should display fetch tool results', async () => {
        render(<FetchContainer/>);

        const urlInput = await screen.getByPlaceholderText('URL', {name: 'url'});
        userEvent.type(urlInput, 'https://jsonplaceholder.typicode.com/todos/1') 

        screen.getByText('GET');
        const methodSelect = await screen.findByLabelText('GET', {name: 'method', id: 'get'})
        userEvent.click(methodSelect);

        const sendButton = await screen.findByRole('button', {name: 'send-button'});
        userEvent.click(sendButton);

        return waitFor( () => {
            const ul = screen.getAllByText('GET', {exact: false});

            expect(ul).toMatchSnapshot();
        })
    })
});