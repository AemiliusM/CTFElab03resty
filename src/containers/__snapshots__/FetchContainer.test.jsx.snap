// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows fetch tool and results should display fetch tool results 1`] = `
Array [
  <label
    for="get"
  >
    GET
  </label>,
  <p>
    GET
  </p>,
]
`;
